<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"></ul></div>

# Домашнее задание

Математический анализ
69 баллов

In [1]:
import pandas as pd
import numpy as np
import sympy as sp
import statsmodels.api as sm
import matplotlib.pyplot as plt
from scipy.optimize import minimize
from scipy.optimize import least_squares
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_percentage_error, r2_score

**Задача 1** (14 баллов)

Найдите экстремумы функции:

$$
f(x, y, z)=2 x^3+2 x y+2 x z+y^2+z^2+2 y-8 .
$$

Распишите подробное решение.

1. Найдем частные производные функции и приравняем их к нулю<br>
1: df/dx = 6x² + 2y + 2z = 0<br>
2: df/dy = 2x + 2y + 2 = 0<br>
3: df/dz = 2x + 2z = 0<br>
<br>
2. Решаем полученную систему уравнений <br>
z = -x (получаем из 3-го)<br>
y = -x - 1 (получаем из 2-го<br>
6x² - 4x - 2 = 0 (получаем из 1-го)<br>
<br>
3. Находим корни квадратного уравнения 6x² - 4x - 2 = 0<br>
(решаем через нахождение дискриминанта)<br>
x₁ = 1, x₂ = -1/3<br>
<br>
4. Находим точку эксремума для x1 = 1 <br>
Подставляем в уравнения в пункте 2, получаем стационарную точку 1<br>
x1,y1,z1 (1; -2; -1)<br>
<br>
5. Находим точку эксремума для x2 = -1/3<br>
Подставляем в уравнения в пункте 2, получаем стационарную точку 2<br>
x2, y2, z2 (-1/3, -2/3, 1/3)<br>
<br>
6. Находим вторые производные:<br>
d2f/dx² = 12x<br>
d2f/dy² = 2<br>
d2f/dz² = 2<br>
d2f/dxdy = 2<br>
d2f/dxdz = 2<br>
d2f/dydz = 0<br>
<br>
7. Строим матрицу Гессе<br>
(12x 2 2)<br>
(2 2 2)<br>
(2 2 0)<br>
<br>
8. Cчитаем определитель матрицы Гессе <br>
12x*2*0+2*2*2+2*2*2-2*2*2-2*2*12x-0*2*2=0+8+8-8-48=8-48x<br>
<br>
9. Считаем определитель матрицы Гессе для стационарных точек (1; -2; -1) и (-1/3, -2/3, 1/3)<br>
det(H1) = -40<br>
det(H2) = 24<br>
<br>
10. Определение характера стационарных точек в соответствии с детерминантом<br><br>
x1,y1,z1 (1; -2; -1)<br>
определитель = -40<br>
отрицательный => в этой точке у функции седловая точка<br>

    x2, y2, z2 (-1/3, -2/3, 1/3)<br>
определитель = 24<br>
Определитель положительный, и так как d2fdx2=12x в этой точке равно 12⋅(−13)=−412⋅(−31​)=−4, что отрицательно, это указывает на локальный максимум.<br>



 


**Задача 2** (25 баллов)

Найдите условные экстремумы функции:

$$
f(x, y)=4 x+8 y, y^2-2 x y+5=0
$$

Вычислите результат самостоятельно (вручную) и с помощью Python. Сравните результаты.

Решаем самостоятельно:

1. Составим функцию множителей Лагранжа<br>
L(x, y, lamda) = 4x+8y+lambda*(y²-2xy+5)
<br>
<br>
2. Находим частные производные <br>
dL/dx = 4-2*lambda*y<br>
dL/dy = 8+2*lambda*y-2*x*lambda<br>
dL/dlambda = 𝑦²-2*x*y+5<br>
<br>

3. Приравниваем производные к 0 и составляем систему уравнений<br>
4-2*lambda*y = 0<br>
8+2*lambda*y-2*x*lambda = 0<br>
𝑦²-2*x*y+5=0
<br>

4. Решаем систему уравнений<br>
1)
4-2*lambda*y = 0 
lambda = 2/y<br>
2)
8+2*lambda*y-2*x*lambda = 0
8+2*2/y*y-2*x*2/y = 0
x = 3/y<br>
3)
𝑦²-2*x*y+5 = 0
y1=1, y2=-1  => x1=3, x2=-3 
<br><br>
Вывод: Условными экстремумами функции являются точки
(1;3),(-1;-3)

In [2]:
# Определяем переменные и множитель Лагранжа
x, y, l = sp.symbols('x y l')

# Задаем целевую функцию
target_function = 4 * x + 8 * y

# Задаем функцию-ограничение
constraint = y**2 - 2 * x * y + 5

# Составляем функцию Лагранжа
lagrange_function = target_function + l * constraint

# Вычисляем частные производные функции Лагранжа
partial_dx = sp.diff(lagrange_function, x)
partial_dy = sp.diff(lagrange_function, y)
partial_dl = sp.diff(lagrange_function, l)

# Решаем систему уравнений, приравнивая производные к нулю
solutions = sp.solve([partial_dx, partial_dy, partial_dl], [x, y, l])

# Выводим точки условного экстремума
print(f'Точки экстремума: ({solutions[0][0]};{solutions[0][1]}), ({solutions[1][0]};{solutions[1][1]})')

Точки экстремума: (-3;-1), (3;1)


Вывод: точки совпадают в обоих решениях

**Задача 3** (30 баллов)

Вам предложены данные с информацией о успеваемости студентов:

Независимые переменные:
* Hours Studied: Общее количество часов, потраченных на учебу каждым студентом.
* Previous Scores: Баллы, полученные студентами на предыдущих экзаменах.
* Sleep Hours: Среднее количество часов сна студента в сутки.
* Sample Question Papers Practiced: Количество пробных экзаменационных работ, с которыми студент занимался.
  
Целевая переменная:
* Performance Index: Показатель общей успеваемости каждого студента. Индекс успеваемости отражает академическую успеваемость студента и округляется до ближайшего целого числа. Индекс варьируется от 10 до 100, при этом более высокие значения свидетельствуют о более высокой успеваемости.

**Решите задачу линейной регрессии, реализовав градиентный спуск самостоятельно, не используя готовое решение из какой-либо библиотеки.**

Скачаем и изучим данные

In [3]:
# посмотрим общую инфу о датасете
df = pd.read_csv(r"C:\Users\Igor\Desktop\study\матан\Student_Performance.txt")
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 6 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Hours Studied                     10000 non-null  int64  
 1   Previous Scores                   10000 non-null  int64  
 2   Extracurricular Activities        10000 non-null  object 
 3   Sleep Hours                       10000 non-null  int64  
 4   Sample Question Papers Practiced  10000 non-null  int64  
 5   Performance Index                 10000 non-null  float64
dtypes: float64(1), int64(4), object(1)
memory usage: 468.9+ KB
None


In [4]:
# посмотрим первые записи
df.head()

Hours Studied  Previous Scores Extracurricular Activities  Sleep Hours  \
0              7               99                        Yes            9   
1              4               82                         No            4   
2              8               51                        Yes            7   
3              5               52                        Yes            5   
4              7               75                         No            8   

   Sample Question Papers Practiced  Performance Index  
0                                 1               91.0  
1                                 2               65.0  
2                                 2               45.0  
3                                 2               36.0  
4                                 5               66.0

Extracurricular Activities с булевым значением нам не понадобится, можно его не использовать

In [5]:
df.describe().T

count     mean        std   min   25%  \
Hours Studied                     10000.0   4.9929   2.589309   1.0   3.0   
Previous Scores                   10000.0  69.4457  17.343152  40.0  54.0   
Sleep Hours                       10000.0   6.5306   1.695863   4.0   5.0   
Sample Question Papers Practiced  10000.0   4.5833   2.867348   0.0   2.0   
Performance Index                 10000.0  55.2248  19.212558  10.0  40.0   

                                   50%   75%    max  
Hours Studied                      5.0   7.0    9.0  
Previous Scores                   69.0  85.0   99.0  
Sleep Hours                        7.0   8.0    9.0  
Sample Question Papers Practiced   5.0   7.0    9.0  
Performance Index                 55.0  71.0  100.0

In [6]:
# посмотрим матрицу корелляции
df.corr()

Hours Studied  Previous Scores  Sleep Hours  \
Hours Studied                          1.000000        -0.012390     0.001245   
Previous Scores                       -0.012390         1.000000     0.005944   
Sleep Hours                            0.001245         0.005944     1.000000   
Sample Question Papers Practiced       0.017463         0.007888     0.003990   
Performance Index                      0.373730         0.915189     0.048106   

                                  Sample Question Papers Practiced  \
Hours Studied                                             0.017463   
Previous Scores                                           0.007888   
Sleep Hours                                               0.003990   
Sample Question Papers Practiced                          1.000000   
Performance Index                                         0.043268   

                                  Performance Index  
Hours Studied                              0.373730  
Previous Scores                            0.915189  
Sleep Hours                                0.048106  
Sample Question Papers Practiced           0.043268  
Performance Index                          1.000000

Данные очищенные, булевый признак использовать не будем. Есть сильная корелляция между Performance Index (целевой признак) и Previous Scores (1 из признаков). Сильной линейной зависимости между оставшимися признаками нет, можно приступить к созданию  градиентного спуска самостоятельно. 

In [7]:
# Разделим данные на признаки и таргет 
X = df[['Hours Studied', 'Previous Scores', 'Sleep Hours', 'Sample Question Papers Practiced']]
y = df[['Performance Index']]

In [8]:
#преобразуем данные в массивы numpy для построения модели линейной регрессии
X = X.values
y = y.values

# Проведем стандартизацию
scaler = StandardScaler()
X = scaler.fit_transform(X)

# Добавим свободный член
X = np.append(np.ones((len(X), 1)), X, axis=1)

In [9]:
# Гиперпараметры
learning_rate = 0.01
num_iterations = 1000

# Инициализация коэффициентов
def initialize_coefficients(n_features):
    return np.zeros((n_features, 1))

# Функция потерь (MSE)
def compute_loss(y, y_pred):
    return (1/(2*len(y))) * np.sum((y - y_pred)**2)

# Градиентный спуск
def gradient_descent(X, y, learning_rate, num_iterations):
    n_samples, n_features = X.shape
    beta = initialize_coefficients(n_features)
    loss_history = []
    
    for iteration in range(num_iterations):
        # Прогноз
        y_pred = np.dot(X, beta)
        # Градиент
        gradient = -(1/n_samples) * np.dot(X.T, (y - y_pred))
        # Обновление коэффициентов
        beta -= learning_rate * gradient
        # Вычисление потерь
        current_loss = compute_loss(y, y_pred)
        loss_history.append(current_loss)
        
        # Вывод потерь каждые 100 итераций
        if iteration % 100 == 0:
            print(f'Iteration {iteration}: Loss = {current_loss:.6f}')
    
    return beta, loss_history

# Разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Обучение модели
beta, loss_history = gradient_descent(X_train, y_train, learning_rate, num_iterations)

# Предсказание на тестовых данных
y_pred = np.dot(X_test, beta)

# Оценка точности модели
mse = compute_loss(y_test, y_pred)
print(f'Среднеквадратичное отклонение (MSE): {mse:.6f}')

# Сравнение с sklearn
sk_model = LinearRegression()
sk_model.fit(X_train[:,1:], y_train)  # без столбца единиц
sk_pred = sk_model.predict(X_test[:,1:])

# Коэффициенты сравнения
print('Коэффициенты модели:')
print(beta.flatten())
print('Коэффициенты модели sklearn')
print(np.append(sk_model.intercept_, sk_model.coef_))

# Сравнение MSE
sk_mse = compute_loss(y_test, sk_pred)
print(f'Среднеквадратичное отклонение модели sk_learn (MSE sklearn): {sk_mse:.6f}')

Iteration 0: Loss = 1714.017500
Iteration 100: Loss = 230.726352
Iteration 200: Loss = 32.700466
Iteration 300: Loss = 6.225091
Iteration 400: Loss = 2.680238
Iteration 500: Loss = 2.204902
Iteration 600: Loss = 2.141067
Iteration 700: Loss = 2.132481
Iteration 800: Loss = 2.131325
Iteration 900: Loss = 2.131169
Среднеквадратичное отклонение (MSE): 2.091015
Коэффициенты модели:
[55.23830455  7.38615123 17.63701614  0.80279009  0.5501177 ]
Коэффициенты модели sklearn
[55.24069363  7.3866529  17.6377233   0.80264289  0.54971847]
Среднеквадратичное отклонение модели sk_learn (MSE sklearn): 2.091127


Вывод: коэффициенты модели идентичные, посчитаем MAPE и коэф. детерминации

In [10]:
mape_our_model = mean_absolute_percentage_error(y_test, y_pred) * 100
print(f'MAPE нашей модели: {mape_our_model:.2f}%')

mape_sk_learn = mean_absolute_percentage_error(y_test, sk_pred) * 100
print(f'MAPE модели sklearn: {mape_our_model:.2f}%')

MAPE нашей модели: 3.54%
MAPE модели sklearn: 3.54%


In [11]:
r2_our_model = r2_score(y_test, y_pred) * 100
print(f'R2 нашей модели: {r2_our_model:.2f}%')

r2_sk_learn = r2_score(y_test, sk_pred) * 100
print(f'R2 модели sklearn: {r2_sk_learn:.2f}%')

R2 нашей модели: 98.87%
R2 модели sklearn: 98.87%


Коэффициент детерминации нашей модели, как и модели sk_learn очень высокий. Модель объясняет 98.87% значений целевой величины.

Ради интереса воспользуемся библиотекой sm сделаем модель линейной регрессии с ее помощью и посмотрим красивый отчет

In [12]:
# Добавление константы (столбец единиц) для свободного члена
X_train_sm = sm.add_constant(X_train[:, 1:])  # Убираем столбец единиц, так как sm.add_constant добавит его

# Создание модели OLS
model_sm = sm.OLS(y_train, X_train_sm)

# Обучение модели
results_sm = model_sm.fit()

# Вывод красивого отчета
print(results_sm.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.988
Model:                            OLS   Adj. R-squared:                  0.988
Method:                 Least Squares   F-statistic:                 1.709e+05
Date:                Sat, 21 Dec 2024   Prob (F-statistic):               0.00
Time:                        00:22:53   Log-Likelihood:                -17151.
No. Observations:                8000   AIC:                         3.431e+04
Df Residuals:                    7995   BIC:                         3.435e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         55.2407      0.023   2392.215      0.0

Тут все тоже очень красиво, коэффициенты примерно такиеже